### PR TITLE
RELATED: RAIL-3382 First version of the FilterBar

### PR DESCRIPTION
**Pull request by no23reason**
Adds first version of the FilterBar component. This needed adding support for the following in the sdk-ui-dashboard package:

* SCSS styling
* SVG handling
* i18n

Also fixes several other issues found along the way:

* show loading state in teh AttributeFilterButton
* make sure AttributeFilterButton respects initial selected items
* fix DashboardView InsightRenderer so that it can recover from Nod Data errors

### Known issues

* The Show all button gets overlapped by the dashboard content, but I believe this will be fixed once the dashboard content is styled properly.

---

Supported PR commands:

| Command                | Description            |
| ---------------------- | ---------------------- |
| `ok to test`           | Re-run standard checks |
| `extended test`        | BackstopJS tests       |
| `extended check sonar` | SonarQube tests        |

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `check-extended` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by gdgate** [comment]
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/3980/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/5610/ : SUCCESS in 16m 03s

**Comment by gdgate** [comment]
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/3986/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/5621/ : SUCCESS in 15m 11s

File: libs/sdk-ui-dashboard/src/dashboard/Dashboard.tsx [hunk 1]
```diff
@@ -171,11 +182,47 @@ export interface IDashboardProps {
...
+                        dispatch(
+                            changeAttributeFilterSelection(
+                                localIdentifier!,
+                                attributeElements,
+                                negativeSelection ? "NOT_IN" : "IN",
+                            ),
+                        );
+                    } else {
+                        invariant(false, "Unknown filter type");
+                    }
```
Comment (lupko) [inline libs/sdk-ui-dashboard/src/dashboard/Dashboard.tsx:223]: i propose to move this into a separate function.. keep the dashboard component small

File: libs/sdk-ui-dashboard/src/filterBar/DashboardDateFilter.tsx [hunk 2]
```diff
@@ -12,15 +15,35 @@ export interface IDashboardDateFilterProps {
...
+
+    /**
+     * A {@link @gooddata/sdk-ui-filters#IDateFilterOptionsByType} configuration of the Date Filter.
+     */
+    dateFilterOptions: IDateFilterOptionsByType;
+
+    /**
+     * Optionally specify custom filter name. Defaults to "Date range" (or its localized equivalent).
+     */
+    customFilterName?: string;
```
Comment (lupko) [inline libs/sdk-ui-dashboard/src/filterBar/DashboardDateFilter.tsx:41]: i propose to wrap these into an object that would encapsulate the config that should be applied on the date filter.

also, contemplating that perhaps that config needs to land already in props of the filter bar as an optional part of the contract. "filter bar should render these date filter and attribute filters. the date filter should have these settings"

File: libs/sdk-ui-dashboard/src/model/commandHandlers/attributeFilter/handler.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+export function* attributeFilterChangeSelectionCommandHandler(
+    _ctx: DashboardContext,
+    cmd: ChangeAttributeFilterSelection,
+) {
+    yield put(
+        filterContextActions.updateAttributeFilterSelection({
+            elements: cmd.payload.elements,
+            filterLocalId: cmd.payload.filterLocalId,
+            negativeSelection: cmd.payload.selectionType === "NOT_IN",
+        }),
```
Comment (lupko) [inline libs/sdk-ui-dashboard/src/model/commandHandlers/attributeFilter/handler.ts:16]: also need to dispatch the event here.

File: libs/sdk-ui-dashboard/src/model/commandHandlers/dateFilter/handler.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        filterContextActions.upsertDateFilter(
+            isAllTime
+                ? { type: "allTime" }
+                : {
+                      type: cmd.payload.type,
+                      granularity: cmd.payload.granularity,
+                      from: cmd.payload.from,
+                      to: cmd.payload.to,
+                  },
+        ),
```
Comment (lupko) [inline libs/sdk-ui-dashboard/src/model/commandHandlers/dateFilter/handler.ts:26]: also need to dispatch the event here.

File: libs/sdk-ui-dashboard/src/model/state/dateFilterConfig/dateFilterConfigConverters.ts [hunk 5]
```diff
@@ -0,0 +1,142 @@
+// (C) 2019-2021 GoodData Corporation
```
Comment (lupko) [inline libs/sdk-ui-dashboard/src/model/state/dateFilterConfig/dateFilterConfigConverters.ts:1]: i suspect that this kind of code would better fit into _staging? there i wanted to concentrate stuff that may part of some models/code that would belong to other packages. for example this.. looks like we can benefit of having it in the sdk-ui-filters?

File: libs/sdk-ui-dashboard/src/model/state/filterContext/filterContextReducers.ts [hunk 6]
```diff
@@ -1,14 +1,111 @@
...
+
+const upsertDateFilter: FilterContextReducer<
+    PayloadAction<
+        | { readonly type: "allTime" }
+        | {
+              readonly type: DateFilterType;
+              readonly granularity: DateFilterGranularity;
+              readonly from?: DateString | number;
+              readonly to?: DateString | number;
+          }
```
Comment (lupko) [inline libs/sdk-ui-dashboard/src/model/state/filterContext/filterContextReducers.ts:49]: can you extract this to a named type pls?

**Comment by lupko** [comment]
few minor notes otherwise LGTM

**Comment by gdgate** [comment]
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

File: libs/sdk-ui-dashboard/src/dashboard/Dashboard.tsx (hunk 1 above)
Comment (no23reason) [reply libs/sdk-ui-dashboard/src/dashboard/Dashboard.tsx]: Moved to a custom hook.

File: libs/sdk-ui-dashboard/src/filterBar/DashboardDateFilter.tsx (hunk 2 above)
Comment (no23reason) [reply libs/sdk-ui-dashboard/src/filterBar/DashboardDateFilter.tsx]: Moved to anothe type. I am not sure that it is the cleanest solution, but let's revisit that later?

File: libs/sdk-ui-dashboard/src/model/state/dateFilterConfig/dateFilterConfigConverters.ts (hunk 5 above)
Comment (no23reason) [reply libs/sdk-ui-dashboard/src/model/state/dateFilterConfig/dateFilterConfigConverters.ts:1]: Moved there

**Comment by gdgate** [comment]
Build in pipeline check aborted.

File: libs/sdk-ui-dashboard/src/model/state/filterContext/filterContextReducers.ts (hunk 6 above)
Comment (no23reason) [reply libs/sdk-ui-dashboard/src/model/state/filterContext/filterContextReducers.ts]: Extracted all the payload types to named types.

**Review by lupko — APPROVED** [review]
LGTM

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by yenkins** [comment]
### Sonar scan result
 - No issue found. Congratulations!

More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR1432

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul-el7/1167/ : FAILURE in 4m 08s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/1441/ : SUCCESS in 29s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/185/ : SUCCESS in 7m 11s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/5638/ : SUCCESS in 14m 44s

**Pull request merged**